server/routes/index.ts

```diff
@@ -1,136 +1,43 @@
-// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
-const limiter = (args?) => rateLimit({
-  windowMs: 1000 * 60,
-  max: 60,
-  keyGenerator: req => req.fingerprint?.hash ?? '',
-  ...args
-})
-
-const cache = new LRUCache({
-  max: 5000,
-  maxSize: 50000,
-  sizeCalculation: (value, key) => {
-    return 1
-  },
-  ttl: 1000 * 60
-
-})
-
-const abbrv = (s: string | object, len: number = 10): string => {
-  let printout = ''
-  if (typeof s !== 'string') {
-    printout = JSON.stringify(s)
-  } else {
-    printout = s
-  }
-  if (printout.length > len) {
-    printout = printout.slice(0, len) + '...' + printout.slice(printout.length - 5)
-  }
-  return printout
-}
-const cached = (ttl?: number) => (req: Request, res: Response, next: NextFunction): void => {
-  const key = `${req.method}|${req.path}|${JSON.stringify(req.query)}|${JSON.stringify(req.body)}`
-  const keyContentType = key + '|header|content-type'
-  const v = cache.get(key)
-  if (v) {
-    const contentType = cache.get(keyContentType)
-    if (contentType) {
-      console.log(`Cache header hit key=[${keyContentType}] value=`, contentType)
-      res.header('content-type', contentType as string)
-    }
-    console.log(`Cache hit key=[${key}] value=`, abbrv(v), typeof v)
-    res.send(v)
-    return
-  } else {
-    // @ts-expect-error wrapper
-    res.__send = res.send
-    res.send = (r) => {
-      console.log(`Cache set key=[${key}] value=`, abbrv(r), typeof r)
-      cache.set(key, r, { ttl: ttl ?? 60 * 1000 })
-      if (res.hasHeader('content-type')) {
-        const h = res.getHeader('content-type')
-        console.log(`Cache header set key=[${keyContentType}] value=`, h)
-        cache.set(keyContentType, h, { ttl: ttl ?? 60 * 1000 })
-      }
-      // @ts-expect-error wrapper
-      return res.__send(r)
-    }
-  }
-  next()
-}
 
 router.get('/health', async (req, res) => {
   console.log('[/health]', JSON.stringify(req.fingerprint))
   res.send('OK').end()
 })
 
-const substackDomainCache = {
-  // uncomment the below for test purpose
-  'localhost:3100': 'polymorpher.substack.com'
-}
-
-const client = buildClient()
-
 router.get('/substack/api/v1/archive',
   limiter(),
+  substackDomainCache,
   async (req, res) => {
-    const host = req.get('host')
-
-    if (!host) {
-      return
-    }
-
-    let substackDomain = substackDomainCache[host]
-
-    if (!substackDomain) {
-      const [subdomain, sld] = host?.split('.')
-
-      substackDomain = await client.getLandingPage(sld, subdomain)
-
-      if (!substackDomain.endsWith('.substack.com')) {
-        throw new Error('Not substack page')
-      }
-
-      substackDomainCache[host] = substackDomain
-    }
-
+    const { substackDomain } = res.locals
     const { data } = await axiosBase.get(`https://${substackDomain}/api/v1/archive`, { params: req.query })
 
-    res.status(200).send(data)
+    res.status(200).set('content-type', 'application/json; charset=utf-8').send(data)
```

Comment:
is /api/v1/archive always responding JSON? You can also just take headers from the axios response https://axios-http.com/docs/res_schema, and pass it over in res

Reply:
Nice catch. Thanks